### PR TITLE
Adding skip_conditions for groups

### LINKED ISSUE #740: None or zero Visitors answer still shows Visitor section complete interstitial
### Expected behaviour
- Visitors complete interstitial only shown when there are more than 0 visitors.

### Actual behaviour
- Visitors complete interstitial is shown even when there are 0 visitors or skipping forward and not answering the number of visitors question.

### Steps to reproduce the behaviour
- Start household questionnaire
- Skip the Household and Accommodation via navigation
- Complete Household and Accommodation section
- Observe that the Visitors complete interstitial is shown incorrectly.

#### Browser
Chrome

#### Operating System
macOS

### Screenshot
N/A

### PR DESCRIPTION
**Pull request by benjefferies**
### What is the context of this PR?
We need to be able to skip the visitors completed interstitial page in
the census household questionnaire. It lives in a separate group to the
visitor questions therefore it won't be skipped by the repeat rule.

This change implements skip_condition on a group to solve that.

Fixes #740 and Fixes
https://trello.com/c/wGkp3T52/291-remove-need-to-update-state-in-navigator

### CODE REVIEW

File: app/data/test_skip_condition_group.json [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                            ]
+                        }
+                    ]
+                }
+            ]
+        },
+        {
+            "id": "should-skip-group",
+            "title": "This question may or may not be skipped",
+            "skip_condition": {
```
Comment (jonnyshaw89) [inline app/data/test_skip_condition_group.json:64]: do you need to add this to schema-v1.json?

File: app/questionnaire/navigator.py [hunk 2]
```diff
@@ -199,8 +191,12 @@ def get_blocks(self):
         blocks = []
 
         for group_index, group in enumerate(SchemaHelper.get_groups(self.survey_json)):
-            no_of_repeats = 1
+            if 'skip_condition' in group:
```
Comment (jonnyshaw89) [inline app/questionnaire/navigator.py:194]: if we are putting this on groups maybe we want to support blocks too? would have made some of the census routing rule a bit simpler?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/data/test_skip_condition_group.json (hunk 1 above)
Comment (benjefferies) [reply app/data/test_skip_condition_group.json:64]: Yep, still a bit WIP but wanted to get some views on how we are implementing this. On the whole it works but it needs tidy ups like your suggestion

File: app/questionnaire/navigator.py (hunk 2 above)
Comment (benjefferies) [reply app/questionnaire/navigator.py:194]: Might make sense one day but let's do it in a feature driven manor then just add it now. Looks at the S3 bucket stuff, isn't used.

File: app/data/test_skip_condition_group.json [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    "questionnaire_id": "0",
+    "schema_version": "0.0.1",
+    "data_version": "0.0.1",
+    "survey_id": "0",
+    "title": "Skip group",
+    "theme": "default",
+    "description": "A questionnaire to skipping a group.",
+    "eq_id": "0",
+    "messages": {
+      "INTEGER_TOO_LARGE": "Number is too large",
```
Comment (collisdigital) [inline app/data/test_skip_condition_group.json]: Do we actually support these at the top-level like this, would have saved copy/pasting them against each answer in the Census schema!

File: app/questionnaire/rules.py [hunk 4]
```diff
@@ -69,6 +69,20 @@ def evaluate_repeat(repeat_rule, answer_store):
         return repeat_function(filtered)
 
 
+def evaluate_skip_condition(skip_condition, answer_store):
```
Comment (collisdigital) [inline app/questionnaire/rules.py]: Pretty sure there is already at least one if not more of these identical blocks of code about parsing/evaluating when rules, can it be consolidated?

Comment (benjefferies) [reply app/questionnaire/rules.py]: There are now 3, it should be possible to consolidate. I started doing it initially and it was a bit of a can of worms so planned on coming back to it later so thanks for the reminder 👍

File: app/data/test_skip_condition_group.json (hunk 3 above)
Comment (benjefferies) [reply app/data/test_skip_condition_group.json]: Will take a look to see if we can remove, we need to spend some time making sure the schema is correct and introduce validation so we wouldn't have to cross reference against the code

Comment (benjefferies) [reply app/data/test_skip_condition_group.json]: Done

File: app/data/test_skip_condition_group.json (hunk 1 above)
Comment (benjefferies) [reply app/data/test_skip_condition_group.json:64]: Done

File: app/questionnaire/rules.py (hunk 4 above)
Comment (iwootten) [reply app/questionnaire/rules.py]: I think this method name should be more generic. Perhaps something like evaluate_when_list, given it doesn't just apply to skip_conditions, it's equally applicable to the when attribute of any rule.

Comment (benjefferies) [reply app/questionnaire/rules.py]: Done

File: app/data/schema/schema-v1.json [hunk 5]
```diff
@@ -176,12 +188,16 @@
                               }
                             },
                             "skip_condition": {
```
Comment (jonnyshaw89) [inline app/data/schema/schema-v1.json:190]: is this correct? arn't skip conditions on sections?

File: tests/functional/pages/surveys/skip_condition_group/do-you-want-to-skip.page.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// >>> WARNING THIS PAGE WAS AUTO-GENERATED ON 2016-12-19 12:17:40.003901 - DO NOT EDIT!!! <<<
```
Comment (jonnyshaw89) [inline tests/functional/pages/surveys/skip_condition_group/do-you-want-to-skip.page.js]: these still have timestamps, you can regenerate without

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/data/schema/schema-v1.json (hunk 5 above)
Comment (benjefferies) [reply app/data/schema/schema-v1.json:190]: Who knows they appear on both in the state items!

Comment (benjefferies) [reply app/data/schema/schema-v1.json:190]: Looks to be on questions

**Review by jonnyshaw89 — APPROVED** [review]

File: app/questionnaire/rules.py [hunk 7]
```diff
@@ -69,6 +53,35 @@ def evaluate_repeat(repeat_rule, answer_store):
...
+    """
+    for when_rule in when_rules:
+        if 'id' in when_rule:
+            answer_index = when_rule['id']
+            filtered = answer_store.filter(answer_id=answer_index, group_instance=group_instance)
+
+            assert len(filtered) <= 1, "Condition will not met: Multiple ({:d}) answers found".format(len(filtered))
+
+            if len(filtered) == 1 and not evaluate_rule(when_rule, filtered[0]['value']):
+                return evaluate_rule(when_rule, filtered[0]['value'])
```
Comment (iwootten) [inline app/questionnaire/rules.py]: Can't you just return False, given this is part of the if?

**Review by iwootten — CHANGES_REQUESTED** [review]

File: app/questionnaire/rules.py (hunk 7 above)
Comment (iwootten) [reply app/questionnaire/rules.py]: In fact, think that's what it originally did right?

Comment (benjefferies) [reply app/questionnaire/rules.py]: Fixed and rebase to speed up process of merging

**Review by iwootten — APPROVED** [review]

**Pull request merged**